File: src/devices/rooms.ts
/** Controlled vocabulary of household rooms (ADR-006). */

export const DEVICE_ROOM_IDS = [
  "master_bedroom",
  "master_bathroom",
  "kitchen_living",
  "garage",
  "office",
  "poker_room",
  "play_room",
  "kids_room",
  "kids_office",
  "guest_room",
] as const;

export type DeviceRoomId = (typeof DEVICE_ROOM_IDS)[number];

export type DeviceRoomDef = {
  id: DeviceRoomId;
  labelRu: string;
  /** Lowercase aliases including the id itself. */
  aliases: readonly string[];
};

export const DEVICE_ROOMS: readonly DeviceRoomDef[] = [
  {
    id: "master_bedroom",
    labelRu: "Спальня",
    aliases: ["master_bedroom", "master bedroom", "спальня"],
  },
  {
    id: "master_bathroom",
    labelRu: "Ванная",
    aliases: ["master_bathroom", "master bathroom", "ванная"],
  },
  {
    id: "kitchen_living",
    labelRu: "Кухня",
    aliases: [
      "kitchen_living",
      "kitchen",
      "living room",
      "livingroom",
      "кухня",
    ],
  },
  {
    id: "garage",
    labelRu: "Гараж",
    aliases: ["garage", "гараж"],
  },
  {
    id: "office",
    labelRu: "Офис",
    aliases: ["office", "офис", "кабинет"],
  },
  {
    id: "poker_room",
    labelRu: "Покерная комната",
    aliases: ["poker_room", "poker room", "покерная", "покерная комната"],
  },
  {
    id: "play_room",
    labelRu: "Игровая",
    aliases: ["play_room", "play room", "playroom", "игровая"],
  },
  {
    id: "kids_room",
    labelRu: "Детская (Василиса)",
    aliases: [
      "kids_room",
      "kid's room",
      "kids room",
      "kidsroom",
      "vasilisa's room",
      "vasilisas room",
      "комната василисы",
      "детская",
      "детская комната",
    ],
  },
  {
    id: "kids_office",
    labelRu: "Детский офис (Василиса)",
    aliases: [
      "kids_office",
      "kid's office",
      "kids office",
      "vasilisa's office",
      "vasilisas office",
      "василисин офис",
      "василисин оффис",
      "детский офис",
    ],
  },
  {
    id: "guest_room",
    labelRu: "Гостевая",
    aliases: ["guest_room", "guest room", "guestroom", "гостевая"],
  },
];

function normalizeKey(raw: string): string {
  return raw
    .trim()
    .toLowerCase()
    .replace(/ё/g, "е")
    .replace(/\s+/g, " ");
}

const ALIAS_TO_ID: Map<string, DeviceRoomId> = (() => {
  const map = new Map<string, DeviceRoomId>();
  for (const room of DEVICE_ROOMS) {
    for (const alias of room.aliases) {
      map.set(normalizeKey(alias), room.id);
    }
  }
  return map;
})();

const LABEL_BY_ID: Record<DeviceRoomId, string> = Object.fromEntries(
  DEVICE_ROOMS.map((r) => [r.id, r.labelRu]),
) as Record<DeviceRoomId, string>;

/** Map id or any alias → stable room id. Empty → undefined. Unknown → null. */
export function normalizeRoom(
  input: string | undefined | null,
): DeviceRoomId | undefined | null {
  if (input == null) return undefined;
  const key = normalizeKey(input);
  if (!key) return undefined;
  return ALIAS_TO_ID.get(key) ?? null;
}

export function roomLabelRu(id: DeviceRoomId): string {
  return LABEL_BY_ID[id];
}

export function isDeviceRoomId(value: string): value is DeviceRoomId {
  return (DEVICE_ROOM_IDS as readonly string[]).includes(value);
}

File: clients/desktop/jarvis_client/rooms.py
"""Controlled vocabulary of household rooms (mirrors Core ADR-006)."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class RoomDef:
    id: str
    label_ru: str
    aliases: tuple[str, ...]


DEVICE_ROOMS: tuple[RoomDef, ...] = (
    RoomDef(
        "master_bedroom",
        "Спальня",
        ("master_bedroom", "master bedroom", "спальня"),
    ),
    RoomDef(
        "master_bathroom",
        "Ванная",
        ("master_bathroom", "master bathroom", "ванная"),
    ),
    RoomDef(
        "kitchen_living",
        "Кухня",
        ("kitchen_living", "kitchen", "living room", "livingroom", "кухня"),
    ),
    RoomDef("garage", "Гараж", ("garage", "гараж")),
    RoomDef("office", "Офис", ("office", "офис", "кабинет")),
    RoomDef(
        "poker_room",
        "Покерная комната",
        ("poker_room", "poker room", "покерная", "покерная комната"),
    ),
    RoomDef(
        "play_room",
        "Игровая",
        ("play_room", "play room", "playroom", "игровая"),
    ),
    RoomDef(
        "kids_room",
        "Детская (Василиса)",
        (
            "kids_room",
            "kid's room",
            "kids room",
            "kidsroom",
            "vasilisa's room",
            "vasilisas room",
            "комната василисы",
            "детская",
            "детская комната",
        ),
    ),
    RoomDef(
        "kids_office",
        "Детский офис (Василиса)",
        (
            "kids_office",
            "kid's office",
            "kids office",
            "vasilisa's office",
            "vasilisas office",
            "василисин офис",
            "василисин оффис",
            "детский офис",
        ),
    ),
    RoomDef(
        "guest_room",
        "Гостевая",
        ("guest_room", "guest room", "guestroom", "гостевая"),
    ),
)


def _norm(raw: str) -> str:
    return " ".join(raw.strip().lower().replace("ё", "е").split())


_ALIAS_TO_ID: dict[str, str] = {}
for _room in DEVICE_ROOMS:
    for _alias in _room.aliases:
        _ALIAS_TO_ID[_norm(_alias)] = _room.id

_LABEL_BY_ID: dict[str, str] = {r.id: r.label_ru for r in DEVICE_ROOMS}


def normalize_room(value: str | None) -> str | None:
    """Return catalog id, or None if empty/unknown."""
    if value is None:
        return None
    key = _norm(value)
    if not key:
        return None
    return _ALIAS_TO_ID.get(key)


def room_label_ru(room_id: str) -> str:
    return _LABEL_BY_ID.get(room_id, room_id)


def room_choices() -> list[tuple[str, str]]:
    """(id, label_ru) for Settings combo."""
    return [(r.id, r.label_ru) for r in DEVICE_ROOMS]
